### PR TITLE
Add linux armv7l and win/osx arm64 builds

### LINKED ISSUE #2104: Freetube - Armhf
### Guidelines

- [x] I have searched the [issue tracker](https://github.com/FreeTubeApp/FreeTube/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

Is not currently build for raspberry pi Armhf

### Proposed Solution

Build a version for armhf

### Alternatives Considered

Flatkpak does not work

### Issue Labels

new feature

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by Jai-JAP**
---
Add linux armv7l and win/osx arm64 builds
---

**Pull Request Type**
Please select what type of pull request this is:
- [ ] Bugfix
- [x] Feature Implementation

**Related issue**
Closes #2104 

**Description**
Added linux armv7l to build.yml and release.yml 
Added win/osx arm64 to _scripts/build.js, build.yml and release.yml 
Linux armv7l and win/osx arm64 artifacts are automatically built with every commit & also releases with every FreeTube release

**Screenshots (if appropriate)**
Not applicable

**Testing (for code that is not small enough to be easily understandable)**
Has this pull request been tested? No
Please describe shortly how you tested it and whether there are any ramifications remaining. 
This PR has not been tested as workflows can't be run on forks

**Desktop (please complete the following information):**
 - OS: Linux armv7l / Windows arm64 / OSX arm64
 - OS Version: N/A
 - FreeTube version: N/A

### CODE REVIEW

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
i think this also fixes https://github.com/FreeTubeApp/FreeTube/issues/1372

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
@PrestonN i think ur the only one that is able to test this because u have a pi

**Comment by Jai-JAP** [comment]
The upoming commits will add windows and macos arm64 builds to releases and every commit ci builds along with required fixes in _scripts/build.js

**Comment by Jai-JAP** [comment]
this is now ready to merge @efb4f5ff-1298-471a-8973-3d47447115dc @PrestonN

**Comment by absidue** [comment]
@Jai-JAP please fix the merge conflicts

**Comment by Jai-JAP** [comment]
Any updates?

File: _scripts/build.js [hunk 1]
```diff
@@ -10,19 +10,23 @@ let targets
 const platform = os.platform()
 const cpus = os.cpus()
 
-if (platform === 'darwin') {
+if (platform == 'darwin') {
```
Comment (ChunkyProgrammer) [inline _scripts/build.js:13]:
```suggestion
if (platform === 'darwin') {
```

File: _scripts/build.js [hunk 2]
```diff
@@ -10,19 +10,23 @@ let targets
...
-  targets = Platform.MAC.createTarget(['dmg'], arch)
-} else if (platform === 'win32') {
-  targets = Platform.WINDOWS.createTarget()
-} else if (platform === 'linux') {
+  if (args[2] === 'arm64') {
+    arch = Arch.arm64
+  }
+  
+  targets = Platform.WINDOWS.createTarget(['nsis', 'zip', 'portable', 'squirrel'], arch)
+} else if (platform == 'linux') {
```
Comment (ChunkyProgrammer) [inline _scripts/build.js:29]:
```suggestion
} else if (platform === 'linux') {
```

File: _scripts/build.js [hunk 3]
```diff
@@ -10,19 +10,23 @@ let targets
...
 
-// Macbook Air 2020 with M1 = 'Apple M1'
-  // Macbook Pro 2021 with M1 Pro = 'Apple M1 Pro'
-  if (cpus[0].model.startsWith('Apple')) {
+  if (args[2] === 'arm64') {
     arch = Arch.arm64
   }
+  
+  targets = Platform.MAC.createTarget(['DMG','zip'], arch)
+} else if (platform == 'win32') {
```
Comment (ChunkyProgrammer) [inline _scripts/build.js:21]:
```suggestion
} else if (platform === 'win32') {
```

**Review by ChunkyProgrammer — COMMENTED** [review]
Looking at the PR, i only notice a small change (use strict equals)

**Comment by Jai-JAP** [comment]
> Looking at the PR, i only notice a small change (use strict equals)

Done

**Comment by PikachuEXE** [comment]
So... how do we test the build output (the compiled apps)
I have M1 Macbooks to test osx-arm64 but can't test the others

**Comment by Jai-JAP** [comment]
If a build is triggered i can test linux armv7l and arm64. But idk about win-arm64 builds

**Comment by PikachuEXE** [comment]
I just trigger build on my own forked repo to test
e.g. https://github.com/PikachuEXE/FreeTube/commit/ada2a76bea2f65dbc228b34e0538d2e09dd9017d

**Comment by Jai-JAP** [comment]
@PikachuEXE See https://github.com/Jai-JAP/FreeTube/actions/runs/2546746799 for builds

**Comment by Jai-JAP** [comment]
Fixed https://github.com/Jai-JAP/FreeTube/actions/runs/2546873058

**Comment by PikachuEXE** [comment]
Trying to test `freetube-0.16.0-mac-arm64.dmg` from
https://github.com/Jai-JAP/FreeTube/actions/runs/2546873058

Got corrupted app
image

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -6,26 +6,36 @@ name: Build
 on:
   push:
     branches: [ master, development, '**-RC' ]
+  workflow_dispatch:
```
Comment (PikachuEXE) [inline .github/workflows/build.yml:9]: Remove empty section

**Review by PikachuEXE — CHANGES_REQUESTED** [review]

**Comment by PikachuEXE** [comment]
I tried using latest `development` with your branch
https://github.com/PikachuEXE/FreeTube/actions/runs/2546964804

Same issue

**Comment by Jai-JAP** [comment]
> Trying to test `freetube-0.16.0-mac-arm64.dmg` from https://github.com/Jai-JAP/FreeTube/actions/runs/2546873058
> 
> Got corrupted app image

that is due to dmg not being signed

File: .github/workflows/build.yml (hunk 4 above)
Comment (Jai-JAP) [reply .github/workflows/build.yml:9]: thats only for testing purposes.
i'll remove that once all builds are tested

**Comment by PikachuEXE** [comment]
if I use latest `development`, update `_scripts/build.js` to your version, run `npm run build:arm64`
image

**Comment by Jai-JAP** [comment]
I got arm64 and armv7l linux builds to work without any issues.

**Comment by Jai-JAP** [comment]
See https://github.com/Jai-JAP/FreeTube/actions/runs/2547106231 for macos zip artifacts

**Comment by PikachuEXE** [comment]
Same issue
Maybe GH Action not suitable for building arm64...?

**Comment by Jai-JAP** [comment]
GH action ia suitable for arm64 builds as many other projects do use them for compiling macos arm64 builds

**Comment by PikachuEXE** [comment]
I retried my another build 
https://github.com/PikachuEXE/FreeTube/actions/runs/2548122796

Which is a branch with `development` as based, plus this branch merged
image

However I can use apps built locally (via `yarn run build` / `yarn run build:arm64`)

And I notice that the file sizes on GH build are smaller than what I build locally
Screenshot 2022-06-24 at 09 53 10
Screenshot 2022-06-24 at 09 53 14
image

**Comment by PikachuEXE** [comment]
Comparison of build task log on my local machine vs GH Action run ([2548122796](https://github.com/PikachuEXE/FreeTube/actions/runs/2548122796))

image

image
image

**Comment by PikachuEXE** [comment]
Just tried the zip files since I have only tried DMG files before
But still x64 works, arm64 doesn't

image

**Comment by Jai-JAP** [comment]
Looks like cross compiling is an issue with electron 16 since i know that this goes well with other projects with electron >= 17.x.y

**Comment by Jai-JAP** [comment]
Can u try a build with electron 17 on gh actions

**Comment by PikachuEXE** [comment]
PR for electron 17.x: https://github.com/FreeTubeApp/FreeTube/pull/2358
Build with electron 17.x: https://github.com/PikachuEXE/FreeTube/actions/runs/2559583515

Same result: (GH) x64 works, arm64 doesn't
I can even try electron 18 but takes more time to study its breaking changes

**Comment by Jai-JAP** [comment]
Can you try
`sudo xattr -r -d com.apple.quarantine "/Applications/*.app"`

**Comment by Jai-JAP** [comment]
See https://github.com/th-ch/youtube-music/issues/614#issuecomment-1047699964

**Comment by PikachuEXE** [comment]
Tested https://github.com/PikachuEXE/FreeTube/actions/runs/2548122796 arm64 version with 
`sudo xattr -r -d com.apple.quarantine` and it seems working
But how to solve this automatically if possible?
And how come x64 version don't have this issue?

**Comment by Jai-JAP** [comment]
IDK but this is an issue only with unsigned apps on arm64 i.e; M1 macs

Also is the dmg working now???

**Comment by PikachuEXE** [comment]
There is no difference for arm64 version between zip/dmg version:
Both requires `sudo xattr -r -d com.apple.quarantine` on `.app` to make it able to be launched
This issue makes MacOS arm64 version build pretty annoying to be used
I would prefer having the build for MacOS arm64 disabled until issue is resolved (I can just build my own version locally)
Normal MacOS users can still use the x64 version without issue (just not native)

**Comment by Jai-JAP** [comment]
For this to be fixed freetube builds need to be signed atleast for mac

**Comment by Jai-JAP** [comment]
Mac OS arm64 builds are now disabled.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -168,30 +213,72 @@ jobs:
...
+      with:
+        name: freetube-${{ steps.versionNumber.outputs.result }}-mac-x64.dmg
         path: build/freetube-${{ steps.versionNumber.outputs.result }}.dmg
+        
+    - name: Upload Mac arm64 .dmg Artifact
+      uses: actions/upload-artifact@v2
+      if: startsWith(matrix.os, 'macos') && startsWith(matrix.runtime, 'osx-arm64')
+      with:
+        name: freetube-${{ steps.versionNumber.outputs.result }}-mac-arm64.dmg
+        path: build/freetube-${{ steps.versionNumber.outputs.result }}-arm64.dmg
```
Comment (PikachuEXE) [inline .github/workflows/build.yml:270]: Still need to be disabled
Also why not put MacOS arm64 entries together?

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -152,13 +201,80 @@ jobs:
...
+    - name: Upload Mac arm64 .dmg Release
+      uses: actions/upload-release-asset@v1
+      if: startsWith(matrix.os, 'macos') && startsWith(matrix.runtime, 'osx-arm64')
+      env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      with:
+        upload_url: https://uploads.github.com/repos/FreeTubeApp/FreeTube/releases/${{ secrets.UPLOAD_ID }}/assets{?name,label}
+        asset_name: freetube-${{ env.PACKAGE_VERSION }}-mac-arm64.dmg
+        asset_path: build/freetube-${{ env.PACKAGE_VERSION }}-arm64.dmg
+        asset_content_type: application/x-apple-diskimage
```
Comment (PikachuEXE) [inline .github/workflows/release.yml:257]: Same issue as another line comment

**Review by PikachuEXE — CHANGES_REQUESTED** [review]

**Comment by Jai-JAP** [comment]
Mac OS arm64 dmg builds are now disabled.

New GH action build artifacts [here](https://github.com/Jai-JAP/FreeTube/actions/runs/2562910411)

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -6,26 +6,37 @@ name: Build
...
+  workflow_dispatch:
 
 jobs:
   build:
     strategy:
       matrix:
         node-version: [16.x]
-        runtime: [ linux-x64, linux-arm64, win-x64, osx-x64 ]
+#        runtime: [ linux-x64, linux-armv7l, linux-arm64, win-x64, win-arm64, osx-x64, osx-arm64 ]
+        runtime: [ linux-x64, linux-armv7l, linux-arm64, win-x64, win-arm64, osx-x64 ]
```
Comment (PikachuEXE) [inline .github/workflows/build.yml:17]:
```suggestion
        runtime: 
          - linux-x64
          - linux-armv7l
          - linux-arm64
          - win-x64
          - win-arm64
          - osx-x64
          # `osx-arm64` disabled due to "macOS gatekeeper"
          # See details in https://github.com/FreeTubeApp/FreeTube/pull/2113
          # - osx-arm64
```

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -13,19 +13,29 @@ jobs:
     strategy:
       matrix:
         node-version: [16.x]
-        runtime: [ linux-x64, linux-arm64, win-x64, osx-x64 ]
+#        runtime: [ linux-x64, linux-armv7l, linux-arm64, win-x64, win-arm64, osx-x64, osx-arm64 ]
+        runtime: [ linux-x64, linux-armv7l, linux-arm64, win-x64, win-arm64, osx-x64 ]
```
Comment (PikachuEXE) [inline .github/workflows/release.yml:17]:
```suggestion
        runtime: 
          - linux-x64
          - linux-armv7l
          - linux-arm64
          - win-x64
          - win-arm64
          - osx-x64
          # `osx-arm64` disabled due to "macOS gatekeeper"
          # See details in https://github.com/FreeTubeApp/FreeTube/pull/2113
          # - osx-arm64
```

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -152,13 +201,80 @@ jobs:
...
+#    - name: Upload Mac arm64 .dmg Release
+#      uses: actions/upload-release-asset@v1
+#      if: startsWith(matrix.os, 'macos') && startsWith(matrix.runtime, 'osx-arm64')
+#      env:
+#          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+#      with:
+#        upload_url: https://uploads.github.com/repos/FreeTubeApp/FreeTube/releases/${{ secrets.UPLOAD_ID }}/assets{?name,label}
+#        asset_name: freetube-${{ env.PACKAGE_VERSION }}-mac-arm64.dmg
+#        asset_path: build/freetube-${{ env.PACKAGE_VERSION }}-arm64.dmg
+#        asset_content_type: application/x-apple-diskimage
```
Comment (PikachuEXE) [inline .github/workflows/release.yml:257]: My bad
This doesn't have to be commented due to `startsWith(matrix.runtime, 'osx-arm64')`
But it's also OK to stay commented, no need to update

**Review by PikachuEXE — CHANGES_REQUESTED** [review]
I hate reading long horizontal line...

**Comment by PikachuEXE** [comment]
I think I am done after this requested change
I can't test arm64 on windows / linux so I will rely on other reviewers / "production test"

File: .github/workflows/release.yml (hunk 9 above)
Comment (Jai-JAP) [reply .github/workflows/release.yml:265]: I'll leave this as-is. As it can be enabled easily and will prevent unnecessary action log entries

**Comment by Jai-JAP** [comment]
> I think I am done after this requested change I can't test arm64 on windows / linux so I will rely on other reviewers / "production test"

Linux armv7l and arm64 are tested working. The only thing that needs tedting is win-arm64

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
@Jai-JAP doesnt this also fix #1372?

**Comment by Jai-JAP** [comment]
I have seen this issue before...
That is an error caused by user not using correct steps for compiling and not an issue with freetube.

Seeing the log contents says that builder has downloaded fpm-linux-x64 on arm linux and hence cannot execute error

**Comment by Jai-JAP** [comment]
@klem-everywhere this is not an issue for webcord

**Comment by Jai-JAP** [comment]
Any updates??

**Review by PikachuEXE — APPROVED** [review]
I can only test MacOS but we know it GGed
Not sure if got people to test other 2...

**Comment by Jai-JAP** [comment]
> I can only test MacOS but we know it GGed
> Not sure if got people to test other 2...

I can test win-x64 and linux-all

**Comment by ChunkyProgrammer** [comment]
i think the workflow passing is good enough for now as the team does not have the hardware to adequately test all of the new builds, if any issues arise from these builds then it's likely an issue caused by electron as the build workflow is passing

**Review by ChunkyProgrammer — APPROVED** [review]

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]

**Comment by apandada1** [comment]
I have a armhf Tablet running RPiOS. I can test this. Where are the build files for armhf?

**Comment by Jai-JAP** [comment]
https://github.com/FreeTubeApp/FreeTube/actions/runs/3003447109
You can download arm nightly builds here.

**Comment by apandada1** [comment]
I can confirm that FreeTube works (I can watch videos, subscribe to channels and so on) on my Surface RT tablet running armhf version of Raspberry Pi OS. While the video is a bit choppy at 780p (the NVidia graphics of Surface RT is not yet supported on GNU/Linux), FreeTube works very well in 360p. At least, now I can peacefully use YouTube on this tablet.

Thank you all for the effort.
image

**Comment by apandada1** [comment]
Is it possible to also enable arm32 builds for Windows? Then one can use the app in Surface RT (running Windows 8.1 RT)

**Comment by Jai-JAP** [comment]
No its not possible because nodejs doesn't support windows arm32 platform.

**Pull request merged**